opm/material/fluidsystems/GenericFluidSystem.hh

```diff
@@ -0,0 +1,252 @@
+// -*- mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+// vi: set et ts=4 sw=4 sts=4:
+/*
+  Copyright 2023 SINTEF Digital, Mathematics and Cybernetics.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+
+#ifndef OPM_GENERICFLUIDSYSTEM_HH
+#define OPM_GENERICFLUIDSYSTEM_HH
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+
+#include <opm/material/fluidsystems/BaseFluidSystem.hpp>
+#include <opm/material/fluidsystems/PTFlashParameterCache.hpp> // TODO: this is something else need to check
+#include <opm/material/viscositymodels/LBC.hpp>
+
+#include <cassert>
+#include <string>
+#include <string_view>
+
+#include <fmt/format.h>
+
+namespace Opm {
+
+    template <typename Scalar>
+    struct ComponentParam {
+        std::string name;
+        Scalar molar_mass;
+        Scalar critic_temp;
+        Scalar critic_pres;
+        Scalar critic_vol;
+        Scalar acentric_factor;
+
+        ComponentParam(const std::string_view name_, const Scalar molar_mass_, const Scalar critic_temp_,
+                       const Scalar critic_pres_, const Scalar critic_vol_, const Scalar acentric_factor_)
+                       : name(name_),
+                         molar_mass(molar_mass_),
+                         critic_temp(critic_temp_),
+                         critic_pres(critic_pres_),
+                         critic_vol(critic_vol_),
+                         acentric_factor(acentric_factor_)
+        {}
+    };
+/*!
+ * \ingroup FluidSystem
+ *
+ * \brief A two phase system that can contain NumComp components
+ *
+ * \tparam Scalar  The floating-point type that specifies the precision of the numerical operations.
+ * \tparam NumComp The number of the components in the fluid system.
+ */
+    template<class Scalar, int NumComp>
+    class GenericFluidSystem : public BaseFluidSystem<Scalar, GenericFluidSystem<Scalar, NumComp> > {
+    public:
+        // TODO: I do not think these should be constant in fluidsystem, will try to make it non-constant later
+        static const int numPhases=2;
```

Comment:
The number of phases are still hard-coded to be two phase, maybe we should change the name to be `GenericTwoPhaseFluidSystem` or `GenericOilGasFluidSystem`. 

Not sure we will have a really generic fluid system that the phases are also arbitrary soon.